vita3k/modules/SceAvPlayer/SceAvPlayer.cpp

```diff
@@ -163,17 +159,102 @@ static Ptr<uint8_t> get_buffer(const PlayerPtr &player, MediaType media_type,
+        {
+            const std::lock_guard<std::mutex> lock(callback_thread->mutex);
+            callback_thread->to_do = ThreadToDo::exit;
+        }
+    };
+    return new_thread_id;
+}
+
+void destroy_callback_thread(HostState &host, SceUID callback_thread_id) {
+    if (callback_thread_id != SCE_KERNEL_ERROR_ILLEGAL_THREAD_ID) {
+        host.kernel.threads.erase(callback_thread_id);
+    }
+}
+
+SceUID create_avplayer_callback_thread(HostState &host, SceUID base_thread_id) {
+    return create_callback_thread(host, base_thread_id, "AvPlayerCallbackThread");
+}
+
+int run_callback(HostState &host, SceUID callback_thread_id, Address callback_address, uint arglen, uint32_t args[]) {
+    if (callback_thread_id == SCE_KERNEL_ERROR_ILLEGAL_THREAD_ID) {
+        LOG_ERROR("Callback thread error");
+        return 0;
+    }
+    const ThreadStatePtr thread = util::find(callback_thread_id, host.kernel.threads);
+    std::unique_lock<std::mutex> lock(thread->mutex);
+    stop(*thread->cpu);
+    assert(arglen <= 4);
+    for (int i = 0; i < arglen; i++) {
+        write_reg(*thread->cpu, i, args[i]);
+    }
+    // TODO: remaining arguments should be pushed into stack
+    write_pc(*thread->cpu, callback_address);
+    lock.unlock();
+    run_thread(*thread, true);
+    return read_reg(*thread->cpu, 0);
+}
+
+uint run_event_callback(HostState &host, SceUID thread_id, const PlayerPtr player_info, uint32_t event_id, uint32_t source_id, Ptr<void> event_data) {
+    constexpr char *export_name = "SceAvPlayerEventCallback";
+    if (player_info->event_manager.event_callback) {
+        SceUID temp_callback_thread;
+        if (player_info->callback_depth_counter > 0) {
+            temp_callback_thread = create_callback_thread(host, thread_id, "AvPlayerCallbackThread_2");
+        } else {
+            temp_callback_thread = player_info->player_callback_thread_id;
+        }
+        uint32_t argp_arr[4] = { player_info->event_manager.user_data, event_id, source_id, event_data.address() };
+        player_info->callback_depth_counter++;
+        auto result = run_callback(host, temp_callback_thread, player_info->event_manager.event_callback.address(), 4, argp_arr);
+        player_info->callback_depth_counter--;
+        if (player_info->callback_depth_counter > 0) {
+            destroy_callback_thread(host, temp_callback_thread);
+        }
+        return result;
+    } else {
+        if (event_id == SCE_AVPLAYER_STATE_READY) {
+            return UNIMPLEMENTED();
+        } else {
+            return 0;
+        }
+    }
+}
+//end of callback_thread
+
 EXPORT(int, sceAvPlayerAddSource, SceUID player_handle, const char *path) {
     const PlayerPtr &player_info = lock_and_find(player_handle, host.kernel.players, host.kernel.mutex);
 
     player_info->player.queue(expand_path(host.io, path, host.pref_path));
-
+    run_event_callback(host, thread_id, player_info, SCE_AVPLAYER_STATE_BUFFERING, 0, Ptr<void>(0)); //may be important for sound
```

Comment:
This will not work unfortunately. If you call run_callback twice, only first call will work correctly, and second call will be ignored. This is because of weird bug in our callback implementation. Refer to https://github.com/Vita3K/Vita3K/issues/809.

Reply:
Oh I just realized that you're running callback on another thread. That should work fine.

Reply:
Now, I see what you're trying to do. We can actually move these callback related code to actual kernel module. Your idea's interesting. It doesn't look thread safe though. You can probably maintain a "job queue" for run_callback and run them one by one on the callback thread. run_callback can wait for callback call to be finished.